extensions/pdh/PDHCounters.cpp

```diff
@@ -0,0 +1,132 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PDHCounters.h"
+#include "utils/StringUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+DWORD PDHCounterBase::getDWFormat() const {
+  return is_double_format_ ? PDH_FMT_DOUBLE : PDH_FMT_LARGE;
+}
+
+PDHCounterBase* PDHCounterBase::createPDHCounter(const std::string& query_name, bool is_double) {
+  auto groups = utils::StringUtils::split(query_name, "\\");
+  if (groups.size() != 2 || query_name.substr(0, 1) != "\\")
+    return nullptr;
+  if (query_name.find("(*)") != std::string::npos) {
+    return new PDHCounterArray(query_name, is_double);
+  }  else {
+    return new PDHCounter(query_name, is_double);
+  }
+}
+
+const std::string& PDHCounterBase::getName() const {
+  return pdh_english_counter_name_;
+}
+
+std::string PDHCounterBase::getObjectName() const {
+  auto groups = utils::StringUtils::split(pdh_english_counter_name_, "\\");
+  return groups[0];
+}
+
+std::string PDHCounterBase::getCounterName() const {
+  auto groups = utils::StringUtils::split(pdh_english_counter_name_, "\\");
+  return groups[1];
+}
+
+void PDHCounter::addToJson(rapidjson::Value& body, rapidjson::Document::AllocatorType& alloc) const {
+  rapidjson::Value key(getCounterName().c_str(), getCounterName().length(), alloc);
+  rapidjson::Value& group_node = acquireNode(getObjectName(), body, alloc);
+  group_node.AddMember(key, getValue(), alloc);
+}
+
+PDH_STATUS PDHCounter::addToQuery(PDH_HQUERY& pdh_query)  {
+  return PdhAddEnglishCounter(pdh_query, pdh_english_counter_name_.c_str(), NULL, &counter_);
+}
+
+PDH_STATUS PDHCounter::collectData() {
+  return PdhGetFormattedCounterValue(counter_, getDWFormat(), NULL, &current_value_);
+}
```

Comment:
You could use `nullptr` instead of `NULL` here, and in other usages below

Reply:
yeah, good idea I changed NULL to nullptr in cases when it was ptr type and into 0 when the argument was integer (e.g. PdhAddEnglishCounterA after many typedefs wants an unsigned __int64).